File: src/Billing/Api/Internal/MaxRateHandler.cs
using Billing.Config.TenantConfigSync.Api;
using Billing.Mediation.Model;
using Grpc.Core;
using Telcobright.Billing.V1;
using MaxRateEngine = Billing.Mediation.Rating.MaxRateEngine;
using MedFacts = Billing.Mediation.Rating.CallFacts;
using MedServiceType = Billing.Mediation.Rating.ServiceType;
using TierInput = Billing.Mediation.Rating.TierInput;

namespace Billing.Service.Api.Internal;

/// <summary>
/// Implements <c>GetMaxRatePerMinute</c> (internal-by-convention; <see cref="BillingServiceImpl"/> delegates
/// here). Maps the proto request to engine facts, resolves the entry tenant's ancestor chain into per-tier
/// rater inputs, runs <see cref="MaxRateEngine"/>, and maps the result back to the proto reply.
/// </summary>
public sealed class MaxRateHandler
{
    private readonly MaxRateEngine _engine;
    private readonly ITenantRegistry _registry;
    private readonly ILogger<MaxRateHandler> _log;

    public MaxRateHandler(MaxRateEngine engine, ITenantRegistry registry, ILogger<MaxRateHandler> log)
    {
        _engine = engine;
        _registry = registry;
        _log = log;
    }

    public Task<MaxRateReply> Handle(MaxRateRequest request, ServerCallContext context)
    {
        var facts = new MedFacts(request.Tenant, request.PartnerId, request.CallingNumber,
            request.CalledNumber, request.SourceIp, MapServiceType(request.ServiceType), request.StartEpochMillis);

        var result = _engine.Resolve(facts, BuildChain(facts, request.Levels));

        _log.LogInformation(
            "GetMaxRatePerMinute tenant={Tenant} partner={Partner} {Caller}->{Called} tiers={Tiers} ok={Ok}",
            request.Tenant, request.PartnerId, request.CallingNumber, request.CalledNumber,
            result.Tiers.Count, result.Ok);

        var reply = new MaxRateReply { Ok = result.Ok, RejectReason = result.RejectReason };
        foreach (var (dbName, tier) in result.Tiers)
        {
            var tr = new TierResult
            {
                DbName = tier.DbName,
                PartnerId = tier.PartnerId,
                ServiceGroup = tier.ServiceGroupId,
                RejectReason = tier.RejectReason,
            };
            foreach (var c in tier.Candidates)
                tr.Candidates.Add(new RateCandidate
                {
                    PackageAccountId = c.PackageAccountId,
                    Uom = c.Uom,
                    RatePerMinute = c.RatePerMinute,
                    MaxAmountFirstMinute = c.MaxAmountFirstMinute,
                });
            reply.Tiers.Add(dbName, tr);
        }
        return Task.FromResult(reply);
    }

    /// <summary>Resolve the entry tenant's ancestor chain (leaf→root) and project each tier into a rater
    /// input — the WHOLE chain in one call. The per-tier partner comes from routesphere's <c>levels</c>
    /// (by depth: 0=admin/root … leaf=deepest reseller); absent a level, the leaf falls back to the entry
    /// partner and ancestors to 0.</summary>
    private List<TierInput> BuildChain(MedFacts facts, IEnumerable<Level> levels)
    {
        var levelByDepth = new Dictionary<int, Level>();
        foreach (var lvl in levels) levelByDepth[lvl.Depth] = lvl;

        var chain = _registry.AncestorChain(facts.Tenant);
        var inputs = new List<TierInput>(chain.Count);
        for (var i = 0; i < chain.Count; i++)
        {
            var tenant = chain[i];
            var depth = chain.Count - 1 - i;   // chain[0]=leaf=deepest reseller; chain[last]=root=admin (depth 0)
            var partnerId = levelByDepth.TryGetValue(depth, out var level)
                ? level.PartnerId
                : (i == 0 ? facts.PartnerId : 0);
            var packages = tenant.Context.PartnerIdWisePackageAccounts.TryGetValue((long)partnerId, out var pkgs)
                ? pkgs
                : (IReadOnlyList<PackageAccount>)Array.Empty<PackageAccount>();
            inputs.Add(new TierInput(tenant.DbName, partnerId, tenant.Context.MediationContext, packages, tenant.Context.Partners));
        }
        return inputs;
    }

    private static MedServiceType MapServiceType(ServiceType t) =>
        t == ServiceType.Sms ? MedServiceType.Sms : MedServiceType.Voice;
}

File: java/src/main/java/com/telcobright/billing/api/internal/MaxRateHandler.java
package com.telcobright.billing.api.internal;

import com.telcobright.billing.grpc.Level;
import com.telcobright.billing.grpc.MaxRateReply;
import com.telcobright.billing.grpc.MaxRateRequest;
import com.telcobright.billing.grpc.TierResult;
import com.telcobright.billing.mediation.model.PackageAccount;
import com.telcobright.billing.mediation.rating.CallFacts;
import com.telcobright.billing.mediation.rating.MaxRateEngine;
import com.telcobright.billing.mediation.rating.MaxRateResult;
import com.telcobright.billing.mediation.rating.RateCandidate;
import com.telcobright.billing.mediation.rating.ServiceType;
import com.telcobright.billing.mediation.rating.TierInput;
import com.telcobright.billing.mediation.rating.TierRating;
import com.telcobright.billing.tenantconfigsync.api.ITenantRegistry;
import com.telcobright.billing.tenantconfigsync.model.Tenant;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.jboss.logging.Logger;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Implements {@code GetMaxRatePerMinute} (internal-by-convention; {@code BillingServiceImpl} delegates here).
 * Maps the proto request to engine facts, resolves the entry tenant's ancestor chain into per-tier rater
 * inputs, runs {@link MaxRateEngine}, and maps the result back to the proto reply.
 */
@Singleton
public class MaxRateHandler {
    private static final Logger log = Logger.getLogger(MaxRateHandler.class);

    private final MaxRateEngine _engine;
    private final ITenantRegistry _registry;

    @Inject
    public MaxRateHandler(MaxRateEngine engine, ITenantRegistry registry) {
        this._engine = engine;
        this._registry = registry;
    }

    public MaxRateReply Handle(MaxRateRequest request) {
        CallFacts facts = new CallFacts(request.getTenant(), request.getPartnerId(), request.getCallingNumber(),
                request.getCalledNumber(), request.getSourceIp(), MapServiceType(request.getServiceType()),
                request.getStartEpochMillis());

        MaxRateResult result = _engine.Resolve(facts, BuildChain(facts, request.getLevelsList()));

        log.infof("GetMaxRatePerMinute tenant=%s partner=%d %s->%s tiers=%d ok=%s",
                request.getTenant(), request.getPartnerId(), request.getCallingNumber(), request.getCalledNumber(),
                result.Tiers().size(), result.Ok());

        MaxRateReply.Builder reply = MaxRateReply.newBuilder().setOk(result.Ok()).setRejectReason(result.RejectReason());
        for (Map.Entry<String, TierRating> e : result.Tiers().entrySet()) {
            TierRating tier = e.getValue();
            TierResult.Builder tr = TierResult.newBuilder()
                    .setDbName(tier.DbName())
                    .setPartnerId(tier.PartnerId())
                    .setServiceGroup(tier.ServiceGroupId())
                    .setRejectReason(tier.RejectReason());
            for (RateCandidate c : tier.Candidates())
                tr.addCandidates(com.telcobright.billing.grpc.RateCandidate.newBuilder()
                        .setPackageAccountId(c.PackageAccountId())
                        .setUom(c.Uom())
                        .setRatePerMinute(c.RatePerMinute())
                        .setMaxAmountFirstMinute(c.MaxAmountFirstMinute())
                        .build());
            reply.putTiers(e.getKey(), tr.build());
        }
        return reply.build();
    }

    /** Resolve the entry tenant's ancestor chain (leaf->root) and project each tier into a rater input — the
     * WHOLE chain in one call. The per-tier partner comes from routesphere's {@code levels} (by depth:
     * 0=admin/root ... leaf=deepest reseller); absent a level, the leaf falls back to the entry partner and
     * ancestors to 0. */
    private List<TierInput> BuildChain(CallFacts facts, List<Level> levels) {
        Map<Integer, Level> levelByDepth = new HashMap<>();
        for (Level lvl : levels) levelByDepth.put(lvl.getDepth(), lvl);

        List<Tenant> chain = _registry.AncestorChain(facts.Tenant());
        List<TierInput> inputs = new ArrayList<>(chain.size());
        for (int i = 0; i < chain.size(); i++) {
            Tenant tenant = chain.get(i);
            int depth = chain.size() - 1 - i;   // chain[0]=leaf=deepest reseller; chain[last]=root=admin (depth 0)
            Level level = levelByDepth.get(depth);
            int partnerId = level != null ? level.getPartnerId() : (i == 0 ? facts.PartnerId() : 0);
            List<PackageAccount> packages =
                    tenant.Context.PartnerIdWisePackageAccounts.getOrDefault((long) partnerId, List.of());
            inputs.add(new TierInput(tenant.DbName, partnerId, tenant.Context.MediationContext, packages,
                    tenant.Context.Partners));
        }
        return inputs;
    }

    private static ServiceType MapServiceType(com.telcobright.billing.grpc.ServiceType t) {
        return t == com.telcobright.billing.grpc.ServiceType.SMS ? ServiceType.Sms : ServiceType.Voice;
    }
}
